### PR TITLE
Enable querying entirely cold datasources 

### PR DESCRIPTION
**Pull request by findingrish**
Issue: https://github.com/apache/druid/issues/14989

## Problem 
Currently, datasource schema doesn’t include columns from cold segments. This makes it impossible to query entirely cold datasource. 

## Approach 
- Mechanism to backfill schema for cold segments in the metadata database. Note, that this is required only for segments created prior to enabling CentralizedDatasourceSchema feature. 
- Update datasource schema building logic on the Coordinator to include schema from cold segments. 
- Make Brokers aware of entirely cold datasource. 

## Backfill schema for cold segments

Leverage the existing schema backfill flow added as part of `CentralizedDatasourceSchema` feature. Users are supposed to manually load the cold segments by making their replication factor as 1 and once the schema is backfilled (can be verified from the metadata database) they can unload the segment. 

## Handling entirely cold datasource 

The problem with cold datasource is that Broker just doesn’t know about the datasource if none of the segment are available. So, the datasource wouldn’t even appear on the console for querying. 
We need a way for the Brokers to be aware of cold datasource, so that it can fetch its schema from the Coordinator. 

Currently, brokers request schema for available datasources from Coordinator in each refresh cycle. 
Brokers now poll set of used datasources from the Coordinator first and then request their schema from the Coordinator. 

Once Broker has schema for Cold datasources, it will show up in the console and become available for querying. 

## Key changes 
*  `CoordinatorSegmentMetadataCache` 
    * It runs a scheduled thread to fetch used segments and build datasource schema from cold segments. Hot and cold schema is merged when datasource schema is queried. 
* `BrokerSegmentMetadataCache` 
    * The refresh condition is slightly updated, refresh is executed in each cycle if the feature is enabled. 
    * The refresh logic is also updated to poll used datasources from the Coordinator. This way Broker can fetch cold datasource schema. 
     
This PR has:

- [X] been self-reviewed.
   - [ ] using the [concurrency checklist](https://github.com/apache/druid/blob/master/dev/code-review/concurrency.md) (Remove this item if the PR doesn't have any relation to concurrency.)
- [ ] added documentation for new or modified features or behaviors.
- [ ] a release note entry in the PR description.
- [ ] added Javadocs for most classes and all non-trivial methods. Linked related entities via Javadoc links.
- [ ] added or updated version, license, or notice information in [licenses.yaml](https://github.com/apache/druid/blob/master/dev/license.md)
- [X] added comments explaining the "why" and the intent of the code wherever would not be obvious for an unfamiliar reader.
- [X] added unit tests or modified existing tests to cover new code paths, ensuring the threshold for [code coverage](https://github.com/apache/druid/blob/master/dev/code-review/code-coverage.md) is met.
- [ ] added integration tests.
- [X] been tested in a test Druid cluster.

### CODE REVIEW

File: server/src/main/java/org/apache/druid/server/coordinator/SegmentReplicationStatusManager.java [hunk 1]
```diff
@@ -0,0 +1,204 @@
...
+        datasourceToUnavailableSegments.addTo(segment.getDataSource(), 0);
+      } else {
+        datasourceToUnavailableSegments.addTo(segment.getDataSource(), 1);
+      }
+    }
+
+    return datasourceToUnavailableSegments;
+  }
+
+  public Object2IntMap<String> getDatasourceToDeepStorageQueryOnlySegmentCount()
```
Comment (kfaraz) [inline server/src/main/java/org/apache/druid/server/coordinator/SegmentReplicationStatusManager.java:109]: @findingrish , this seems like the only new method that has been added here.
Please remove this new class `SegmentReplicationStatusManager` and move the code back to`DruidCoordinator`.

If a refactor is required, please do it in a separate PR. 
This PR should focus only on the required changes.

**Review by kfaraz — CHANGES_REQUESTED** [review]
This PR has some refactors which should be tackled separately, in order to facilitate a smoother review of the core changes here.

File: server/src/main/java/org/apache/druid/server/coordinator/SegmentReplicationStatusManager.java (hunk 1 above)
Comment (kfaraz) [reply server/src/main/java/org/apache/druid/server/coordinator/SegmentReplicationStatusManager.java:109]: Correction: It seems that this method had already existed too. 
@findingrish , is there any new code in `SegmentReplicationStatusManager`?

Comment (findingrish) [reply server/src/main/java/org/apache/druid/server/coordinator/SegmentReplicationStatusManager.java:109]: @kfaraz There is no new code in `SegmentReplicationStatusManager`. The reason for refactoring was a cyclic dependency between `CoordinatorSegmentMetadataCache` and `DruidCoordinator` while trying to use `DruidCoordinator#getSegmentReplicationFactor`. 

I will raise a separate PR for the refactor.

Comment (kfaraz) [reply server/src/main/java/org/apache/druid/server/coordinator/SegmentReplicationStatusManager.java:109]: Okay. Can you share some more details on how the cyclic dependency is coming into picture?

Comment (findingrish) [reply server/src/main/java/org/apache/druid/server/coordinator/SegmentReplicationStatusManager.java:109]: Currently `DruidCoordinator` has a dependency on `CoordinatorSegmentMetadataCache`, for this patch I need to use `DruidCoordinator#getSegmentReplicationFactor` in `CoordinatorSegmentMetadataCache` which is resulting in cyclic dependency. 

As a solution, I have refactored `DruidCoordinator` to separate out the code which updates `segmentReplicationStatus` and `broadcastSegments`.

Comment (kfaraz) [reply server/src/main/java/org/apache/druid/server/coordinator/SegmentReplicationStatusManager.java:109]: @findingrish , you could just expose a method `updateSegmentReplicationStatus()` on `CoordinatorSegmentMetadataCache`. Call this method from `DruidCoordinator.UpdateReplicationStatus.run()` where we update `broadcastSegments` and `segmentReplicationStatus`.

Let me know if this works for you.

Comment (findingrish) [reply server/src/main/java/org/apache/druid/server/coordinator/SegmentReplicationStatusManager.java:109]: Yeah, this approach would work for me. 
However, it seems bit odd that `DruidCoordinator.UpdateReplicationStatus` has to additionally update state in some other class, ideally the consumer `CoordinatorSegmentMetadataCache` should be pulling this information? 

Is there a reason to avoid the refactor work?

Comment (kfaraz) [reply server/src/main/java/org/apache/druid/server/coordinator/SegmentReplicationStatusManager.java:109]: > Is there a reason to avoid the refactor work?

Yes, the dependencies are already all over the place which makes the code less readable and also complicates testing. A refactor is needed here but it would have to be thought through a little.

> However, it seems bit odd that DruidCoordinator.UpdateReplicationStatus has to additionally update state in some other class,

Not really, you can think of the `DruidCoordinator` (or rather the `UpdateReplicationStatus` duty in this case) as sending a notification to the `CoordinatorSegmentMetadatCache` that the segment replication status has been updated. The `DruidCoordinator` already sends notification to the metadata cache about leadership status, this is another notification in the same vein.

File: server/src/main/java/org/apache/druid/segment/metadata/CoordinatorSegmentMetadataCache.java [hunk 2]
```diff
@@ -419,6 +502,98 @@ private Set<SegmentId> filterSegmentWithCachedSchema(Set<SegmentId> segmentIds)
...
+              log.debug("[%s] merged signature is unchanged.", ds);
+            }
+
+            return new DataSourceInformation(ds, mergedSignature);
+          }
+      );
+    }
+
+    // remove any stale datasource from the map
+    coldSchemaTable.keySet().retainAll(dataSources);
```
Comment (cryptoe) [inline server/src/main/java/org/apache/druid/segment/metadata/CoordinatorSegmentMetadataCache.java:569]: Do we have a test case for this ?

**Review by cryptoe — COMMENTED** [review]
Left some comments.

File: server/src/main/java/org/apache/druid/server/coordinator/SegmentReplicationStatusManager.java (hunk 1 above)
Comment (findingrish) [reply server/src/main/java/org/apache/druid/server/coordinator/SegmentReplicationStatusManager.java:109]: > Yes, the dependencies are already all over the place which makes the code less readable and also complicates testing. A refactor is needed here but it would have to be thought through a little.

Yes, this makes sense. `DruidCoordinator` refactoring would need more thought. 

Thanks for the suggestion, I will update the patch.

File: server/src/main/java/org/apache/druid/segment/metadata/CoordinatorSegmentMetadataCache.java (hunk 2 above)
Comment (findingrish) [reply server/src/main/java/org/apache/druid/segment/metadata/CoordinatorSegmentMetadataCache.java:569]: Yes, in `CoordinatorSegmentMetadataCacheTest#testColdDatasourceSchema_verifyStaleDatasourceRemoved`.

File: server/src/main/java/org/apache/druid/segment/metadata/CoordinatorSegmentMetadataCache.java [hunk 3]
```diff
@@ -419,6 +502,98 @@ private Set<SegmentId> filterSegmentWithCachedSchema(Set<SegmentId> segmentIds)
...
+  }
+
+  private RowSignature mergeHotAndColdSchema(RowSignature hot, RowSignature cold)
```
[the remark this answers is not in the document]
Comment (findingrish) [reply server/src/main/java/org/apache/druid/segment/metadata/CoordinatorSegmentMetadataCache.java:572]: I can refactor this a bit to have a single method for merging the RowSignature.

File: server/src/main/java/org/apache/druid/segment/metadata/CoordinatorSegmentMetadataCache.java [hunk 4]
```diff
@@ -181,6 +220,12 @@ public void onLeaderStart()
     try {
       segmentSchemaBackfillQueue.onLeaderStart();
       cacheExecFuture = cacheExec.submit(this::cacheExecLoop);
+      coldSchemaExecFuture = coldScehmaExec.schedule(
+          this::coldDatasourceSchemaExec,
+          coldSchemaExecPeriodMillis,
```
[the remark this answers is not in the document]
Comment (findingrish) [reply server/src/main/java/org/apache/druid/segment/metadata/CoordinatorSegmentMetadataCache.java:225]: > Do we have any metrics which tell us the performance of these executor service in terms of number of cold segments back filed 

We are not backfilling segment here. It is just looping over the segments, identifying cold segment and building their schema. 
If the datasource schema is updated it is logged.

File: server/src/main/java/org/apache/druid/segment/metadata/CoordinatorSegmentMetadataCache.java [hunk 5]
```diff
@@ -92,14 +114,27 @@ public CoordinatorSegmentMetadataCache(
...
     this.segmentSchemaCache = segmentSchemaCache;
     this.segmentSchemaBackfillQueue = segmentSchemaBackfillQueue;
+    this.sqlSegmentsMetadataManager = sqlSegmentsMetadataManager;
+    this.segmentReplicationStatusManager = segmentReplicationStatusManager;
+    this.coldSchemaExecPeriodMillis =
+        segmentsMetadataManagerConfigSupplier.get().getPollDuration().getMillis();
+    coldScehmaExec = Executors.newSingleThreadScheduledExecutor(
+        new ThreadFactoryBuilder()
+            .setNameFormat("DruidColdSchema-ScheduledExecutor-%d")
+            .setDaemon(true)
```
[the remark this answers is not in the document]
Comment (findingrish) [reply server/src/main/java/org/apache/druid/segment/metadata/CoordinatorSegmentMetadataCache.java:135]: I will update, we don't need a daemon thread here.

File: server/src/test/java/org/apache/druid/segment/metadata/CoordinatorSegmentMetadataCacheTest.java [hunk 6]
```diff
@@ -1721,12 +1787,207 @@ public void testSameSegmentAddedOnMultipleServer() throws InterruptedException,
...
+        new InternalQueryConfig(),
+        new NoopServiceEmitter(),
+        segmentSchemaCache,
+        backFillQueue,
+        sqlSegmentsMetadataManager,
+        segmentsMetadataManagerConfigSupplier
+    );
+
+    SegmentReplicaCount segmentReplicaCount = new SegmentReplicaCount();
+    segmentReplicaCount.setRequired(0, 0);
```
Comment (kfaraz) [inline server/src/test/java/org/apache/druid/segment/metadata/CoordinatorSegmentMetadataCacheTest.java:1839]: Wouldn't the values already be zero in a fresh instance?

Comment (findingrish) [reply server/src/test/java/org/apache/druid/segment/metadata/CoordinatorSegmentMetadataCacheTest.java:1839]: Right, I was trying to be explicit about setting it to `0`, so that it is clear in the test that the given segment is unavailable.

Comment (kfaraz) [reply server/src/test/java/org/apache/druid/segment/metadata/CoordinatorSegmentMetadataCacheTest.java:1839]: You can just add a comment to that effect and/or use test method names that clarify that point.
Invoking this method requires making it `public` which doesn't really seem necessary since it is only going to be used in this test.

File: server/src/main/java/org/apache/druid/segment/metadata/AbstractSegmentMetadataCache.java [hunk 7]
```diff
@@ -199,8 +199,9 @@ public abstract class AbstractSegmentMetadataCache<T extends DataSourceInformati
   /**
    * Map of datasource and generic object extending DataSourceInformation.
    * This structure can be accessed by {@link #cacheExec} and {@link #callbackExec} threads.
+   * It contains schema for datasources with atleast 1 available segment.
```
Comment (asdf2014) [inline server/src/main/java/org/apache/druid/segment/metadata/AbstractSegmentMetadataCache.java:202]:
```suggestion
   * It contains schema for datasources with at least 1 available segment.
```

**Review by asdf2014 — CHANGES_REQUESTED** [review]

File: server/src/test/java/org/apache/druid/segment/metadata/CoordinatorSegmentMetadataCacheTest.java (hunk 6 above)
Comment (findingrish) [reply server/src/test/java/org/apache/druid/segment/metadata/CoordinatorSegmentMetadataCacheTest.java]: Makes sense. Updated.

File: server/src/main/java/org/apache/druid/segment/metadata/CoordinatorSegmentMetadataCache.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (findingrish) [reply server/src/main/java/org/apache/druid/segment/metadata/CoordinatorSegmentMetadataCache.java:239]: Makes sense. I am logging details if the execution delay is greater than 50 seconds.

File: server/src/main/java/org/apache/druid/client/coordinator/CoordinatorClient.java [hunk 8]
```diff
@@ -69,4 +69,9 @@ public interface CoordinatorClient
    * Returns a new instance backed by a ServiceClient which follows the provided retryPolicy
    */
   CoordinatorClient withRetryPolicy(ServiceRetryPolicy retryPolicy);
+
+  /**
+   * Retrieves list of used datasources.
+   */
+  ListenableFuture<Set<String>> fetchUsedDataSources();
```
[the remark this answers is not in the document]
Comment (findingrish) [reply server/src/main/java/org/apache/druid/client/coordinator/CoordinatorClient.java]: I have updated the method name to `fetchDatasourcesWithUsedSegments` to make it more understandable. 
I don't think we need to document about what `used segments` means, since it is widely referred in the code and docs. For example, here https://druid.apache.org/docs/latest/api-reference/data-management-api/#mark-a-single-segment-as-used.

File: server/src/main/java/org/apache/druid/segment/metadata/CoordinatorSegmentMetadataCache.java [hunk 9]
```diff
@@ -419,6 +521,128 @@ private Set<SegmentId> filterSegmentWithCachedSchema(Set<SegmentId> segmentIds)
...
+            return new DataSourceInformation(ds, mergedSignature);
+          }
+      );
+    }
+
+    // remove any stale datasource from the map
+    coldSchemaTable.keySet().retainAll(dataSources);
+
+    if (stopwatch.millisElapsed() > COLD_SCHEMA_SLOWNESS_THRESHOLD_MILLIS) {
+      log.info("Cold schema processing was slow, taking [%d] millis. "
```
Comment (cryptoe) [inline server/src/main/java/org/apache/druid/segment/metadata/CoordinatorSegmentMetadataCache.java:609]: Else for this should be debug.

File: server/src/main/java/org/apache/druid/segment/metadata/CoordinatorSegmentMetadataCache.java [hunk 10]
```diff
@@ -419,6 +541,94 @@ private Set<SegmentId> filterSegmentWithCachedSchema(Set<SegmentId> segmentIds)
...
+      log.debug("[%s] signature from cold segments is [%s]", dataSourceName, coldSignature);
+
+      coldSchemaTable.put(dataSourceName, new DataSourceInformation(dataSourceName, coldSignature));
+    }
+
+    // remove any stale datasource from the map
+    coldSchemaTable.keySet().retainAll(dataSourceWithColdSegmentSet);
+
+    String executionStatsLog = StringUtils.format(
+        "Cold schema processing was slow, taking [%d] millis. "
```
Comment (cryptoe) [inline server/src/main/java/org/apache/druid/segment/metadata/CoordinatorSegmentMetadataCache.java:610]:
```suggestion
        "Cold schema processing took [%d] millis. "
```

File: server/src/main/java/org/apache/druid/segment/metadata/CoordinatorSegmentMetadataCache.java [hunk 11]
```diff
@@ -419,6 +541,94 @@ private Set<SegmentId> filterSegmentWithCachedSchema(Set<SegmentId> segmentIds)
...
+        + "Processed [%d] datasources, [%d] segments & [%d] datasourceWithColdSegments.",
```
Comment (cryptoe) [inline server/src/main/java/org/apache/druid/segment/metadata/CoordinatorSegmentMetadataCache.java:611]:
```suggestion
        + "Processed total [%d] datasources, [%d] segments. Found [%d] datasources with cold segments.",
```

**Review by cryptoe — APPROVED** [review]
